unix: expand uv_available_parallelism() to support more platforms

oerdnj opened the pull request:
This PR expands the `uv_available_parallelism()` to natively support more platforms:
- FreeBSD
- NetBSD
- OpenBSD + macOS

oerdnj commented:
I have some FreeBSD/OpenBSD/NetBSD VMs lying around, so I'll confirm that everything builds and works fine during the day.

bnoordhuis commented on src/unix/core.c:
```diff
@@ -1869,6 +1879,23 @@ int uv__search_path(const char* prog, char* buf, size_t* buflen) {
 }
 
+#if __NetBSD__
+int CPU_COUNT(cpuset_t *cpuset) {
```
```suggestion
static int CPU_COUNT(cpuset_t *cpuset) {
```
And it'd be better to call it e.g. `uv__cpu_count`

bnoordhuis commented on src/unix/core.c:
```diff
@@ -1906,9 +1933,43 @@ unsigned int uv_available_parallelism(void) {
+  if (0 == cpuset_getaffinity(CPU_LEVEL_WHICH, CPU_WHICH_PID, -1, sizeof(set), &set))
+    rc = CPU_COUNT(&set);
+# elif __NetBSD__
+  cpuset_t *set = cpuset_create();
```
```suggestion
  cpuset_t* set = cpuset_create();
```

bnoordhuis commented on src/unix/core.c:
```diff
@@ -1906,9 +1933,43 @@ unsigned int uv_available_parallelism(void) {
+  };
+
+  for (size_t i = 0; i < ARRAY_SIZE(mib); i++) {
+    if (0 == sysctl(mib[i], ARRAY_SIZE (mib[i]), &nprocs, &len, NULL, 0) &&
```
```suggestion
    if (0 == sysctl(mib[i], ARRAY_SIZE(mib[i]), &nprocs, &len, NULL, 0) &&
```

saghul commented on src/unix/core.c:
```diff
@@ -1906,9 +1933,43 @@ unsigned int uv_available_parallelism(void) {
+#  ifdef HW_NCPUONLINE
+    { CTL_HW, HW_NCPUONLINE },
+#  endif
+    { CTL_HW, HW_NCPU }
```
According to the macOS man page:

```
     hw.ncpu
             The number of cpus. This attribute is deprecated and it is recommended that hw.logicalcpu, hw.logicalcpu_max,
             hw.physicalcpu, or hw.physicalcpu_max be used instead.
```

Not sure if that has been there forever or not.

oerdnj replied:
Hmm, I missed that, but found yet another comment in the `sysctl.h` header:

```
 *   hw.ncpu                   - The maximum number of processors that could be available this boot.
 *                               Use this value for sizing of static per processor arrays; i.e. processor load statistics.
 *
 *   hw.activecpu              - The number of processors currently available for executing threads.
 *                               Use this number to determine the number threads to create in SMP aware applications.
 *                               This number can change when power management modes are changed.
 *
 *   hw.physicalcpu            - The number of physical processors available in the current power management mode.
 *   hw.physicalcpu_max        - The maximum number of physical processors that could be available this boot
 *
 *   hw.logicalcpu             - The number of logical processors available in the current power management mode.
 *   hw.logicalcpu_max         - The maximum number of logical processors that could be available this boot
```
 
I am guessing `hw.activecpu` should be the right one to pick followed by `hw.logicalcpu`?

saghul replied:
Sounds good!

oerdnj replied:
Done; there's a lot of fixup commits (as I fixed the individual platforms) to make the review easier, I will squash them when you say so.

oerdnj commented:
Small test program:
```
#include <uv.h>
#include <stdio.h>

int main(void) {
	printf("Number of available CPUs: %u\n", uv_available_parallelism());
	return 0;
}
```

Result on my macOS:
```
$ gcc -I include/ .libs/libuv.a a.c && a.out
Number of available CPUs: 8
```

oerdnj commented:
`make check` on FreeBSD is broken :-D
```
/usr/local/bin/ld: test/run_tests-test-poll-closesocket.o:/home/ondrej/Projects/libuv/test/test-poll-closesocket.c:29: multiple definition of `handle'; test/run_tests-test-poll-close-doesnt-corrupt-stack.o:/home/ondrej/Projects/libuv/test/test-poll-close-doesnt-corrupt-stack.c:36: first defined here
/usr/local/bin/ld: test/run_tests-test-poll-closesocket.o:/home/ondrej/Projects/libuv/test/test-poll-closesocket.c:28: multiple definition of `sock'; test/run_tests-test-poll-close-doesnt-corrupt-stack.o:/home/ondrej/Projects/libuv/test/test-poll-close-doesnt-corrupt-stack.c:35: first defined here
```

But my test program correctly reports number of CPUs:
```
$ gcc -I include/ -L .libs -luv /tmp/a.c && ./a.out
Number of available CPUs: 6
```

truss output:
```
mmap(0x0,135168,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 43039363956736 (0x2724e2000000)
mprotect(0x13f9eafda000,4096,PROT_READ)		 = 0 (0x0)
issetugid()					 = 0 (0x0)
sigfastblock(0x1,0x13f9eafdcbb0)		 = 0 (0x0)
open("/etc/libmap.conf",O_RDONLY|O_CLOEXEC,034200030030) = 3 (0x3)
fstat(3,{ mode=-rw-r--r-- ,inode=2243636,size=35,blksize=32768 }) = 0 (0x0)
read(3,"includedir /usr/local/etc/libmap"...,35) = 35 (0x23)
close(3)					 = 0 (0x0)
open("/usr/local/etc/libmap.d",O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC,0165) ERR#2 'No such file or directory'
open("/opt/lib/libuv.so.1",O_RDONLY|O_CLOEXEC|O_VERIFY,00) ERR#2 'No such file or directory'
open("/var/run/ld-elf.so.hints",O_RDONLY|O_CLOEXEC,020002024) = 3 (0x3)
read(3,"Ehnt\^A\0\0\0\M^@\0\0\0\M^@\0\0"...,128) = 128 (0x80)
fstat(3,{ mode=-r--r--r-- ,inode=2971613,size=256,blksize=32768 }) = 0 (0x0)
pread(3,"/lib:/usr/lib:/usr/lib/compat:/u"...,128,0x80) = 128 (0x80)
close(3)					 = 0 (0x0)
open("/lib/libuv.so.1",O_RDONLY|O_CLOEXEC|O_VERIFY,00) ERR#2 'No such file or directory'
open("/usr/lib/libuv.so.1",O_RDONLY|O_CLOEXEC|O_VERIFY,00) ERR#2 'No such file or directory'
open("/usr/lib/compat/libuv.so.1",O_RDONLY|O_CLOEXEC|O_VERIFY,00) ERR#2 'No such file or directory'
open("/usr/local/lib/libuv.so.1",O_RDONLY|O_CLOEXEC|O_VERIFY,00) = 3 (0x3)
fstat(3,{ mode=-rwxr-xr-x ,inode=323946,size=179488,blksize=32768 }) = 0 (0x0)
fstatfs(3,{ fstypename=ufs,mntonname=/,mntfromname=/dev/da0p2,fsid= }) = 0 (0x0)
mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_PREFAULT_READ,3,0x0) = 34939105280 (0x822887000)
mmap(0x0,192512,PROT_NONE,MAP_GUARD,-1,0x0)	 = 34920898560 (0x82172a000)
mmap(0x82172a000,69632,PROT_READ,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0x0) = 34920898560 (0x82172a000)
mmap(0x82173b000,110592,PROT_READ|PROT_EXEC,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0x10000) = 34920968192 (0x82173b000)
mmap(0x821756000,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x2a000) = 34921078784 (0x821756000)
mmap(0x821757000,8192,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x2a000) = 34921082880 (0x821757000)
munmap(0x822887000,4096)			 = 0 (0x0)
close(3)					 = 0 (0x0)
open("/opt/lib/libc.so.7",O_RDONLY|O_CLOEXEC|O_VERIFY,00) ERR#2 'No such file or directory'
open("/lib/libc.so.7",O_RDONLY|O_CLOEXEC|O_VERIFY,00) = 3 (0x3)
fstat(3,{ mode=-r--r--r-- ,inode=3685910,size=1957560,blksize=32768 }) = 0 (0x0)
fstatfs(3,{ fstypename=ufs,mntonname=/,mntfromname=/dev/da0p2,fsid= }) = 0 (0x0)
mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_PREFAULT_READ,3,0x0) = 34952548352 (0x823559000)
mmap(0x0,4206592,PROT_NONE,MAP_GUARD,-1,0x0)	 = 34928267264 (0x821e31000)
mmap(0x821e31000,544768,PROT_READ,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0x0) = 34928267264 (0x821e31000)
mmap(0x821eb6000,1355776,PROT_READ|PROT_EXEC,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0x84000) = 34928812032 (0x821eb6000)
mmap(0x822001000,40960,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x1ce000) = 34930167808 (0x822001000)
mmap(0x82200b000,28672,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x1d7000) = 34930208768 (0x82200b000)
mmap(0x822012000,2236416,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_ANON,-1,0x0) = 34930237440 (0x822012000)
munmap(0x823559000,4096)			 = 0 (0x0)
close(3)					 = 0 (0x0)
open("/opt/lib/libthr.so.3",O_RDONLY|O_CLOEXEC|O_VERIFY,00) ERR#2 'No such file or directory'
open("/lib/libthr.so.3",O_RDONLY|O_CLOEXEC|O_VERIFY,00) = 3 (0x3)
fstat(3,{ mode=-r--r--r-- ,inode=3685938,size=128128,blksize=32768 }) = 0 (0x0)
fstatfs(3,{ fstypename=ufs,mntonname=/,mntfromname=/dev/da0p2,fsid= }) = 0 (0x0)
mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_PREFAULT_READ,3,0x0) = 34942959616 (0x822c34000)
mmap(0x0,188416,PROT_NONE,MAP_GUARD,-1,0x0)	 = 34956353536 (0x8238fa000)
mmap(0x8238fa000,53248,PROT_READ,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0x0) = 34956353536 (0x8238fa000)
mmap(0x823907000,73728,PROT_READ|PROT_EXEC,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0xc000) = 34956406784 (0x823907000)
mmap(0x823919000,8192,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x1d000) = 34956480512 (0x823919000)
mmap(0x82391b000,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x1e000) = 34956488704 (0x82391b000)
mmap(0x82391c000,49152,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_ANON,-1,0x0) = 34956492800 (0x82391c000)
munmap(0x822c34000,4096)			 = 0 (0x0)
close(3)					 = 0 (0x0)
mprotect(0x822001000,40960,PROT_READ)		 = 0 (0x0)
mprotect(0x823919000,8192,PROT_READ)		 = 0 (0x0)
mprotect(0x822001000,40960,PROT_READ|PROT_WRITE) = 0 (0x0)
mprotect(0x822001000,40960,PROT_READ)		 = 0 (0x0)
readlink("/etc/malloc.conf",0x820858620,1024)	 ERR#2 'No such file or directory'
issetugid()					 = 0 (0x0)
mmap(0x0,2097152,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON|MAP_ALIGNED(21),-1,0x0) = 43039366053888 (0x2724e2200000)
cap_getmode({ 0 })				 = 0 (0x0)
open("/dev/pvclock",O_RDONLY|O_CLOEXEC,04173053450) = 3 (0x3)
mmap(0x0,192,PROT_READ,MAP_SHARED,3,0x0)	 = 34964930560 (0x824128000)
close(3)					 = 0 (0x0)
mmap(0x0,2097152,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON|MAP_ALIGNED(12),-1,0x0) = 43039368151040 (0x2724e2400000)
mmap(0x0,4194304,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON|MAP_ALIGNED(21),-1,0x0) = 43039370248192 (0x2724e2600000)
thr_self(0x2724e2412000)			 = 0 (0x0)
mmap(0x80085a000,4096,PROT_NONE,MAP_ANON,-1,0x0) = 34947014656 (0x823012000)
rtprio_thread(RTP_LOOKUP,100204,0x820858a6c)	 = 0 (0x0)
sigaction(SIGTHR,{ 0x8239139e0 SA_SIGINFO ss_t },0x0) = 0 (0x0)
sigprocmask(SIG_UNBLOCK,{ },0x0)		 = 0 (0x0)
_umtx_op(0x820858a60,UMTX_OP_WAKE,0x1,0x0,0x0)	 = 0 (0x0)
mprotect(0x0,0,PROT_NONE)			 = 0 (0x0)
thr_wake(0xffffffffffffffff)			 ERR#3 'No such process'
getpid()					 = 4202 (0x106a)
getpid()					 = 4202 (0x106a)
sigprocmask(SIG_BLOCK,{ SIGHUP|SIGINT|SIGQUIT|SIGILL|SIGTRAP|SIGABRT|SIGEMT|SIGFPE|SIGKILL|SIGBUS|SIGSEGV|SIGSYS|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGSTOP|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2 },{ }) = 0 (0x0)
sigfastblock(0x3,0x0)				 = 0 (0x0)
sigprocmask(SIG_SETMASK,{ },0x0)		 = 0 (0x0)
sigfastblock(0x1,0x2724e2412038)		 = 0 (0x0)
getcontext(0x820857f00)				 = 0 (0x0)
sysarch(AMD64_GET_XFPUSTATE,0x820857ec8)	 = 0 (0x0)
fstat(1,{ mode=crw--w---- ,inode=94,size=0,blksize=4096 }) = 0 (0x0)
ioctl(1,TIOCGETA,0x820858c04)			 = 0 (0x0)
Number of available CPUs: 6
write(1,"Number of available CPUs: 6\n",28)	 = 28 (0x1c)
mmap(0x0,135168,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 43039374442496 (0x2724e2a00000)
exit(0x0)
process exit, rval = 0
```

oerdnj commented:
FreeBSD 12.3 results (tests are ok):

```
$ cc -I include/ -L .libs -luv /tmp/a.c && ./a.out
Number of available CPUs: 6
```

truss output:
```
mmap(0x0,131072,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34362011648 (0x80022b000)
mprotect(0x800227000,4096,PROT_READ)		 = 0 (0x0)
issetugid()					 = 0 (0x0)
openat(AT_FDCWD,"/etc/libmap.conf",O_RDONLY|O_CLOEXEC,00) = 3 (0x3)
fstat(3,{ mode=-rw-r--r-- ,inode=2739314,size=122,blksize=32768 }) = 0 (0x0)
read(3,"# $FreeBSD: releng/12.3/libexec/"...,122) = 122 (0x7a)
close(3)					 = 0 (0x0)
open("/usr/local/etc/libmap.d",O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC,0165) ERR#2 'No such file or directory'
openat(AT_FDCWD,".libs/libuv.so.1",O_RDONLY|O_CLOEXEC|O_VERIFY,00) = 3 (0x3)
fstat(3,{ mode=-rwxr-xr-x ,inode=3532394,size=721544,blksize=32768 }) = 0 (0x0)
fstatfs(3,{ fstypename=ufs,mntonname=/,mntfromname=/dev/ada0p2,fsid= }) = 0 (0x0)
mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_PREFAULT_READ,3,0x0) = 34362142720 (0x80024b000)
mmap(0x0,188416,PROT_NONE,MAP_GUARD,-1,0x0)	 = 34362146816 (0x80024c000)
mmap(0x80024c000,69632,PROT_READ,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0x0) = 34362146816 (0x80024c000)
mmap(0x80025d000,106496,PROT_READ|PROT_EXEC,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0x10000) = 34362216448 (0x80025d000)
mmap(0x800277000,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x29000) = 34362322944 (0x800277000)
mmap(0x800278000,8192,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x29000) = 34362327040 (0x800278000)
munmap(0x80024b000,4096)			 = 0 (0x0)
close(3)					 = 0 (0x0)
openat(AT_FDCWD,".libs/libc.so.7",O_RDONLY|O_CLOEXEC|O_VERIFY,00) ERR#2 'No such file or directory'
openat(AT_FDCWD,"/var/run/ld-elf.so.hints",O_RDONLY|O_CLOEXEC,00) = 3 (0x3)
read(3,"Ehnt\^A\0\0\0\M^@\0\0\0k\0\0\0\0"...,128) = 128 (0x80)
fstat(3,{ mode=-r--r--r-- ,inode=2087046,size=235,blksize=32768 }) = 0 (0x0)
pread(3,"/lib:/usr/lib:/usr/lib/compat:/u"...,107,0x80) = 107 (0x6b)
close(3)					 = 0 (0x0)
openat(AT_FDCWD,"/lib/libc.so.7",O_RDONLY|O_CLOEXEC|O_VERIFY,00) = 3 (0x3)
fstat(3,{ mode=-r--r--r-- ,inode=722317,size=1959392,blksize=32768 }) = 0 (0x0)
fstatfs(3,{ fstypename=ufs,mntonname=/,mntfromname=/dev/ada0p2,fsid= }) = 0 (0x0)
mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_PREFAULT_READ,3,0x0) = 34362142720 (0x80024b000)
mmap(0x0,4161536,PROT_NONE,MAP_GUARD,-1,0x0)	 = 34362335232 (0x80027a000)
mmap(0x80027a000,524288,PROT_READ,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0x0) = 34362335232 (0x80027a000)
mmap(0x8002fa000,1314816,PROT_READ|PROT_EXEC,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0x7f000) = 34362859520 (0x8002fa000)
mmap(0x80043b000,40960,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x1bf000) = 34364174336 (0x80043b000)
mmap(0x800445000,28672,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x1c8000) = 34364215296 (0x800445000)
mmap(0x80044c000,2252800,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_ANON,-1,0x0) = 34364243968 (0x80044c000)
munmap(0x80024b000,4096)			 = 0 (0x0)
close(3)					 = 0 (0x0)
openat(AT_FDCWD,".libs/libthr.so.3",O_RDONLY|O_CLOEXEC|O_VERIFY,00) ERR#2 'No such file or directory'
openat(AT_FDCWD,"/lib/libthr.so.3",O_RDONLY|O_CLOEXEC|O_VERIFY,00) = 3 (0x3)
fstat(3,{ mode=-r--r--r-- ,inode=722334,size=131640,blksize=32768 }) = 0 (0x0)
fstatfs(3,{ fstypename=ufs,mntonname=/,mntfromname=/dev/ada0p2,fsid= }) = 0 (0x0)
mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_PREFAULT_READ,3,0x0) = 34362142720 (0x80024b000)
mmap(0x0,184320,PROT_NONE,MAP_GUARD,-1,0x0)	 = 34366496768 (0x800672000)
mmap(0x800672000,53248,PROT_READ,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0x0) = 34366496768 (0x800672000)
mmap(0x80067f000,73728,PROT_READ|PROT_EXEC,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,3,0xc000) = 34366550016 (0x80067f000)
mmap(0x800691000,8192,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x1d000) = 34366623744 (0x800691000)
mmap(0x800693000,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,3,0x1e000) = 34366631936 (0x800693000)
mmap(0x800694000,45056,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_ANON,-1,0x0) = 34366636032 (0x800694000)
munmap(0x80024b000,4096)			 = 0 (0x0)
close(3)					 = 0 (0x0)
mprotect(0x800277000,4096,PROT_READ)		 = 0 (0x0)
mprotect(0x80043b000,36864,PROT_READ)		 = 0 (0x0)
mprotect(0x800691000,4096,PROT_READ)		 = 0 (0x0)
sigprocmask(SIG_BLOCK,{ SIGHUP|SIGINT|SIGQUIT|SIGKILL|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGSTOP|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2 },{ }) = 0 (0x0)
mprotect(0x80043b000,36864,PROT_READ|PROT_WRITE) = 0 (0x0)
sigprocmask(SIG_SETMASK,{ },0x0)		 = 0 (0x0)
sigprocmask(SIG_BLOCK,{ SIGHUP|SIGINT|SIGQUIT|SIGKILL|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGSTOP|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2 },{ }) = 0 (0x0)
sigprocmask(SIG_SETMASK,{ },0x0)		 = 0 (0x0)
sigprocmask(SIG_BLOCK,{ SIGHUP|SIGINT|SIGQUIT|SIGKILL|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGSTOP|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2 },{ }) = 0 (0x0)
mprotect(0x80043b000,36864,PROT_READ)		 = 0 (0x0)
sigprocmask(SIG_SETMASK,{ },0x0)		 = 0 (0x0)
readlink("/etc/malloc.conf",0x7fffffffd470,1024) ERR#2 'No such file or directory'
issetugid()					 = 0 (0x0)
__sysctl("vm.overcommit",2,0x7fffffffd3d4,0x7fffffffd3e0,0x0,0) = 0 (0x0)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34362142720 (0x80024b000)
madvise(0x80024b000,4096,MADV_FREE)		 = 0 (0x0)
munmap(0x80024b000,4096)			 = 0 (0x0)
mmap(0x0,2097152,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366681088 (0x80069f000)
munmap(0x80069f000,2097152)			 = 0 (0x0)
mmap(0x0,4190208,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366681088 (0x80069f000)
munmap(0x80069f000,1445888)			 = 0 (0x0)
munmap(0x800a00000,647168)			 = 0 (0x0)
cap_getmode({ 0 })				 = 0 (0x0)
open("/dev/pvclock",O_RDONLY|O_CLOEXEC,04)	 = 3 (0x3)
mmap(0x0,192,PROT_READ,MAP_SHARED,3,0x0)	 = 34362142720 (0x80024b000)
close(3)					 = 0 (0x0)
mmap(0x0,32768,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366681088 (0x80069f000)
mmap(0x0,4194304,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34370224128 (0x800a00000)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366713856 (0x8006a7000)
mmap(0x0,12288,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366717952 (0x8006a8000)
mmap(0x0,12288,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366730240 (0x8006ab000)
mmap(0x0,12288,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366742528 (0x8006ae000)
sigprocmask(SIG_BLOCK,{ SIGHUP|SIGINT|SIGQUIT|SIGKILL|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGSTOP|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2 },{ }) = 0 (0x0)
sigprocmask(SIG_SETMASK,{ },0x0)		 = 0 (0x0)
__sysctl("kern.usrstack",2,0x80069e6f0,0x7fffffffd8c8,0x0,0) = 0 (0x0)
getrlimit(RLIMIT_STACK,{ cur=536870912,max=536870912 }) = 0 (0x0)
__sysctlbyname("kern.smp.cpus",13,0x80069e5f0,0x7fffffffd8c8,0x0,0) = 0 (0x0)
mmap(0x0,20480,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366754816 (0x8006b1000)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366775296 (0x8006b6000)
thr_self(0x8006b1000)				 = 0 (0x0)
mmap(0x7fffdfffe000,4096,PROT_NONE,MAP_ANON,-1,0x0) = 140736951476224 (0x7fffdfffe000)
rtprio_thread(RTP_LOOKUP,100427,0x7fffffffd888)	 = 0 (0x0)
sigaction(SIGTHR,{ 0x800686240 SA_SIGINFO ss_t },0x0) = 0 (0x0)
sigprocmask(SIG_UNBLOCK,{ },0x0)		 = 0 (0x0)
_umtx_op(0x7fffffffd890,UMTX_OP_WAKE,0x1,0x0,0x0) = 0 (0x0)
mprotect(0x0,0,PROT_NONE)			 = 0 (0x0)
getpid()					 = 9444 (0x24e4)
getpid()					 = 9444 (0x24e4)
sigprocmask(SIG_BLOCK,{ SIGHUP|SIGINT|SIGQUIT|SIGILL|SIGTRAP|SIGABRT|SIGEMT|SIGFPE|SIGKILL|SIGBUS|SIGSEGV|SIGSYS|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGSTOP|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2 },{ }) = 0 (0x0)
sigprocmask(SIG_SETMASK,{ },0x0)		 = 0 (0x0)
getcontext(0x7fffffffcd40)			 = 0 (0x0)
sysarch(AMD64_GET_XFPUSTATE,0x7fffffffcd08)	 = 0 (0x0)
mprotect(0x202000,4096,PROT_READ)		 = 0 (0x0)
cpuset_getaffinity(0x3,0x2,0xffffffffffffffff,0x20,0x7fffffffe800) = 0 (0x0)
fstat(1,{ mode=crw--w---- ,inode=94,size=0,blksize=4096 }) = 0 (0x0)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366779392 (0x8006b7000)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366783488 (0x8006b8000)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366787584 (0x8006b9000)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366791680 (0x8006ba000)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366795776 (0x8006bb000)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366799872 (0x8006bc000)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366803968 (0x8006bd000)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366808064 (0x8006be000)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366812160 (0x8006bf000)
mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366816256 (0x8006c0000)
ioctl(1,TIOCGETA,0x7fffffffe1c0)		 = 0 (0x0)
Number of available CPUs: 6
write(1,"Number of available CPUs: 6\n",28)	 = 28 (0x1c)
mmap(0x0,131072,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34366820352 (0x8006c1000)
exit(0x0)
process exit, rval = 0
```

oerdnj commented:
NetBSD 9.1 results:
```
$ cc -I include/ -L .libs -luv /tmp/a.c && LD_LIBRARY_PATH=.libs ./a.out
  CC       src/unix/libuv_la-core.lo
  CCLD     libuv.la
Number of available CPUs: 6
```

That said, the `sched_getaffinity_np()` always returns empty set on my NetBSD 9.4 machine.

I am going to upgrade the VM to NetBSD 10.0 and try again.

oerdnj commented:
OpenBSD 7.5 works:
```
$ cc -I include/ -L .libs -luv /tmp/a.c && LD_LIBRARY_PATH=.libs ./a.out
Number of available CPUs: 8
```

And since there's no `truss` available, I confirmed the value comes from `HW_NCPUONLINE`.

oerdnj commented:
Turns out that by default the `sched_getaffinity_np()` returns an empty set.  The code is still correct in case the affinity gets set by other means (parent process?) or previous call, so I am going to leave the NetBSD code as it is.

oerdnj commented:
> Turns out that by default the `sched_getaffinity_np()` returns an empty set. The code is still correct in case the affinity gets set by other means (parent process?) or previous call, so I am going to leave the NetBSD code as it is.

There's a chit-chat between NetBSD people also about `pthread_getaffinity_np()`:

> While the behaviour of sched_getaffinity isn't well defined, that's
> not true for the pthread_getaffinity_np function.
> 
> The default affinity mask there is defined as 'all ones' and an
> 'all zeros' mask is invalid for setting affinity.

So, who know what the end result will be - either code update or man page update 🤷.

oerdnj commented:
@bnoordhuis @saghul Anything else needed for the review?  (Apart from squashing the commits into a single one.)

saghul reviewed (approved):
LGTM, nice work @oerdnj !

oerdnj commented:
> LGTM, nice work @oerdnj !

Well, thank you for allowing me to move my #ifdef spaghetti to be your #ifdef spaghetti :-D

Rebased, squashed into a single commit and force pushed.

saghul commented:
The Android failure is unrelated AFAICT.
